bundles/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/json/obj/Value.java

```diff
@@ -0,0 +1,182 @@
+
+import javax.measure.Unit;
+
+import org.eclipse.jdt.annotation.NonNull;
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.tacmi.internal.TACmiBindingConstants;
+import org.openhab.core.library.types.DecimalType;
+import org.openhab.core.library.types.OnOffType;
+import org.openhab.core.library.types.QuantityType;
+import org.openhab.core.library.unit.SIUnits;
+import org.openhab.core.library.unit.Units;
+import org.openhab.core.thing.type.ChannelTypeUID;
+import org.openhab.core.types.State;
+
+/**
+ * Class holding the Value JSON element
+ *
+ * @author Moritz 'Morty' Strübe - Initial contribution
+ */
+@NonNullByDefault
+public class Value {
+    public double value = 0;
+    public String unit = "-1";
+    private @Nullable TypeInfo ti = null;
+
+    private TypeInfo getTi() {
+        // Need a local variable to satisfy null-checks
+        var rv = ti;
+        if (rv != null) {
+            return rv;
+        }
+
+        try {
+            rv = TYPE_MAP.get(Integer.parseInt(unit));
+        } catch (Exception e) {
+        }
+        if (rv == null) {
+            rv = new TypeInfo("Undefined: " + unit, "Number", null);
+        }
+        ti = rv;
+        return rv;
+    }
+
+    public String getDesc() {
+        return getTi().desc;
+    }
+
+    public @Nullable String getType() {
+        return getTi().unitname;
+    }
+
+    public @Nullable ChannelTypeUID getChannelType() {
+        final var el = getTi();
+        if (!el.unitname.startsWith("Number")) {
+            return TACmiBindingConstants.CHANNEL_TYPE_SCHEME_STATE_RO_UID;
+        }
+        return TACmiBindingConstants.CHANNEL_TYPE_SCHEME_NUMERIC_RO_UID;
+    }
+
+    public State getState() {
+        final var el = getTi();
+        if (el.unitname.startsWith("Switch")) {
+            return (value != 0) ? OnOffType.ON : OnOffType.OFF;
+        }
+        if (el.unit != null) {
+            // Must copy to a local variable to make the null-checker happy
+            final var tmp = Objects.requireNonNull(el.unit);
+            return new QuantityType<>(value * el.factor, tmp);
+        }
+        return new DecimalType(Double.valueOf(value * el.factor));
+    }
+
+    static class TypeInfo {
+        public final String desc;
+        public final String unitname;
+        public final @Nullable Unit<@NonNull ?> unit;
+        public final double factor;
+
+        public TypeInfo(String desc, String unitname, @Nullable Unit<@NonNull ?> unit) {
+            this.desc = desc;
+            this.unitname = unitname;
+            this.factor = 1;
+            this.unit = unit;
+        }
+
+        public TypeInfo(String desc, String unitname, double factor, Unit<?> unit) {
+            this.desc = desc;
+            this.unitname = unitname;
+            this.factor = factor;
+            this.unit = unit;
+        }
+    }
+
+    static final Map<Integer, TypeInfo> TYPE_MAP;
```

Comment:
There are more types in this map that can make use of units of measre by adding a Dimension. There are also some types defined with a Dimension, but lack a unit, please provide them.

Reply:
I did my best. I hope I got things right. Units seem a bit complicated and even the original code seems to be a bit inconsistent. :-/

Reply:
Hi @cmorty ,

I had a quick look at your PR and came across this unit mapping. I had something similar for the 'schema' binding. During one of my last PR's I was able to remove the mapping and use OpenHAB's internal QuantityType mapping for most of them. It works fine and should also be able to handle new units when TA adds them.
This code is currently used for unit mapping:
https://github.com/openhab/openhab-addons/blob/main/bundles/org.openhab.binding.tacmi/src/main/java/org/openhab/binding/tacmi/internal/schema/ApiPageParser.java#L382
Maybe it makse sense to to refactor this to a helper/utils class, so both variants could use the same code? The units should be the same as both are from TA ;)

What do you think?

Bye,
    Chris

Reply:
Hmm. Not sure how much that helps after I did all the mapping manually, as I need a mapping from integer (as String!) to the type anyway. I also assume that approach only works for already registered units thus we need to add units for the other types anyway (which was most of the work).

But yes, having a shared unit-name-as-string to unit mapping might be helpful. That way JSON only needs an id-as-string to unit-name-as-string mapping. And thanks to that mapping it should be pretty easy to write a test-case that makes sure all those units are mapped successfully. Or keep the map and use it for the Schema-API, too, to ensure that the mapping is correct and not rely on some magic mapping provided by OpenHab.

Reply:
As the implementation can be changed in a future PR, i mark this as resolved.